src/PoissonFreeSpace/test_poisson.cpp

```diff
@@ -0,0 +1,266 @@
+		auto const &phiExact = ref.array(iter);
+		auto const ncomp = ref.nComp();
+		AMREX_ASSERT(ncomp == 1);
+
+		amrex::ParallelFor(indexRange, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+			// compute exact phi
+			Real const x = prob_lo[0] + (i + Real(0.5)) * dx[0];
+			Real const y = prob_lo[1] + (j + Real(0.5)) * dx[1];
+			Real const z = prob_lo[2] + (k + Real(0.5)) * dx[2];
+			Real const r = std::sqrt(std::pow(x - X0, 2) + std::pow(y - Y0, 2) + std::pow(z - Z0, 2)) / R0;
+
+			Real phi = NAN;
+			if (r < 1.0) {
+				phi = std::pow(r, 6) / 42. - std::pow(r, 7) / 14. + std::pow(r, 8) / 12. - 2. * std::pow(r, 9) / 45. + std::pow(r, 10) / 110. -
+				      (1. / 1260.);
+			} else {
+				phi = -1.0 / (2310. * r);
+			}
+			phiExact(i, j, k) = (R0 * R0) * phi;
+		});
+	}
+}
+
+template <> void RadhydroSimulation<PoissonProblem>::ErrorEst(int lev, amrex::TagBoxArray &tags, Real /*time*/, int /*ngrow*/)
+{
+	// tag cells for refinement
+	Real const jeans_threshold = 0.25;
+
+	auto dx = Geom(lev).CellSizeArray();
+	Real const dx_min = std::min({AMREX_D_DECL(dx[0], dx[1], dx[2])});
+
+	for (amrex::MFIter mfi(state_new_[lev]); mfi.isValid(); ++mfi) {
+		const amrex::Box &box = mfi.validbox();
+		const auto state = state_new_[lev].const_array(mfi);
+		const auto tag = tags.array(mfi);
+
+		amrex::ParallelFor(box, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+			Real const rho = state(i, j, k, HydroSystem<PoissonProblem>::density_index);
+			Real const cs = 1.0 / (4. * 1024. * M_PI); // arbitrary for this test problem
+			// Real const cs = 1.0 / (1024. * M_PI); // arbitrary for this test
+			// problem
+			Real const lambda_jeans = cs * std::sqrt(M_PI / (C::Gconst * rho));
+			Real const jeans_number = dx_min / lambda_jeans;
+
+			if (jeans_number > jeans_threshold && rho > 0.0) {
+				tag(i, j, k) = amrex::TagBox::SET;
+			}
+		});
+	}
+}
+
+auto problem_main() -> int
+{
+	const int nvars = RadhydroSimulation<PoissonProblem>::nvarTotal_;
+	amrex::Vector<amrex::BCRec> boundaryConditions(nvars);
+	for (int n = 0; n < nvars; ++n) {
+		for (int i = 0; i < AMREX_SPACEDIM; ++i) {
+			boundaryConditions[n].setLo(i, amrex::BCType::foextrap);
+			boundaryConditions[n].setHi(i, amrex::BCType::foextrap);
+		}
+	}
+
+	// Problem initialization
+	RadhydroSimulation<PoissonProblem> sim(boundaryConditions);
+	sim.is_hydro_enabled_ = true;
+	sim.is_radiation_enabled_ = false;
+	sim.plotfileInterval_ = -1;
+	sim.checkpointInterval_ = -1;
+
+	// initialize
+	sim.setInitialConditions();
+
+	// set physical boundary conditions (periodic or free space)
+	amrex::BCRec phys_bc;
+	for (int i = 0; i < AMREX_SPACEDIM; ++i) {
+		phys_bc.setLo(i, amrex::PhysBCType::outflow);
+		phys_bc.setHi(i, amrex::PhysBCType::outflow);
+	}
+
+	// solve Poisson equation
+	amrex::Print() << "Initializing gravity solver..." << std::endl;
+
+	Gravity<PoissonProblem> grav(&sim, phys_bc, sim.coordCenter_, HydroSystem<PoissonProblem>::density_index);
```

Comment:
Here I created a `Gravity<PoissonProblem>` object for testing purposes. For real usage, it should all be contained within AMRSimulation/RadhydroSimulation.